### PR TITLE
Refactor walking to make Strategizer walking use recipe walking

### PR DESCRIPTION
**Pull request by raulverag**
Also, RecipeWalker is now a subclass of Walker, which contains the basic machinery for walking.

This makes Walking and RecipeWalking common facilities, which will be used to do recipe resolution in the Common module without a Strategizer or using Strategies, in a subsequent PR.

I've added Maria and Piotr as reviewers since I messed with strategizer and all the strategies.

### CODE REVIEW

**Comment by raulverag** [comment]
Oh dear. I thought I did my changes on a clean new branch, but it seems we've done conflicting changes to walker. Someone combined Walker and WalkerBase, whereas I rearranged WalkerBase, called it Walker, and renamed the old Walker to RecipeWalker. So this will take a bit to disentangle. Please hold off reviewing until I sort this out.

**Comment by raulverag** [comment]
Ok. I think this is cleaned up. All tests pass on my machine. Review away.

File: src/runtime/recipe/walker.ts [hunk 1]
```diff
@@ -5,126 +5,211 @@
...
+    walker.onAction(action);
+    results.forEach(result => {
+      walker.onResult(result);
+      walker.onResultDone();
+    });
+    walker.onActionDone();
+    return walker.descendants;
+  }
+
+  _runUpdateList(recipe, updateList) {
```
Comment (piotrswigon) [inline src/runtime/recipe/walker.ts]: Would you care about renaming the arguments and variables below to not mention _recipes_, as walker is supposed to be a generic base class?

File: src/planning/strategizer.ts [hunk 2]
```diff
@@ -220,71 +223,29 @@ export class Strategizer {
...
-      valid,
-    });
+export class StrategizerWalker extends RecipeWalker {
+  constructor(tactic) {
+    super(tactic);
   }
 
-  onResultDone(): void {
-    this.currentResult = undefined;
+  createDescendant(result, score,): void {
```
Comment (piotrswigon) [inline src/planning/strategizer.ts]: What's up with the comma after _score_? I just learned they are valid, but do you want one here?

Comment (raulverag) [reply src/planning/strategizer.ts]: Just a copy-and-paste typo. Thanks for spotting it.

File: src/runtime/recipe/walker.ts (hunk 1 above)
Comment (raulverag) [reply src/runtime/recipe/walker.ts]: Good point. This required a bit more refactoring, so PTAL.

File: src/runtime/recipe/recipe-walker.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  onResult(result) {
+    super.onResult(result);
+    const recipe: Recipe = result.result;
+    const updateList = [];
+
+    // update phase - walk through recipe and call onRecipe,
+    // onHandle, etc.
+
+    if (this.onRecipe) {
+      result = this.onRecipe(recipe, result);
```
Comment (piotrswigon) [inline src/runtime/recipe/recipe-walker.ts]: Do you want to override the argument or create a new variable? In below lines you're creating new variables, here you override the argument. I assume you want the former, but the naming clash is unfortunate.

I'm thinking whether we could rename `onResult`/`onResultDone` to something like `onItem`/`onItemDone`. From perspective of this method, the `result` is something that will be created by applying the update function to the item passed in. WDYT?

File: src/runtime/recipe/recipe-walker.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        }
+      }
+    }
+
+    this._runUpdateList(recipe, updateList);
+  }
+
+  createDescendant(recipe, score): void {
+    const valid = recipe.normalize();
+    //if (!valid) debugger;
```
Comment (piotrswigon) [inline src/runtime/recipe/recipe-walker.ts]: clean up?

**Pull request merged**